react-renderer/src/inspector/views/Mod.tsx

```diff
@@ -0,0 +1,161 @@
+import * as React from "react";
+import IViewProps from "./IViewProps";
+import AttrPicker from "./mod/AttrPicker";
+import { staticMap } from "src/componentMap";
+import styled from "styled-components";
+import defmap from "./mod/defmap";
+
+const ShapeItem = styled.li<any>`
+  display: block;
+  padding: 1em;
+  margin-top: -1px;
+  border: 1px solid #d1d1d1;
+  background-color: ${({ selected }: any) =>
+    selected ? "#F9F9F9" : "#f0f0f0"};
+  color: rgba(0, 0, 0, 0.5);
+  font-family: monospace;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  cursor: pointer;
+`;
+
+interface IState {
+  selectedShape: number;
+}
+
+class Mod extends React.Component<IViewProps, IState> {
+  public readonly state = { selectedShape: -1};
+  public setSelectedShape = (key: number) => {
+    this.setState({ selectedShape: key});
+  };
+  public modAttr = (attrname: string, attrval: Value<any>) => { 
+    const {frame, modShapes, frameIndex, history} = this.props;
+    const {selectedShape} = this.state;
+    if (frameIndex === -1 || frameIndex === history.length - 1) { // make it only work on last frame
+      const shap = frame!.shapes[(selectedShape === -1 ? 0 : selectedShape)]; // the -1 index behaves weirdly so workaround
+      if (shap.properties.hasOwnProperty(attrname)) {
+        shap.properties[attrname] = attrval;
```

Comment:
This seems prone to bugs because you're not supposed to mutate properties in-place if they're from props. Props should always be immutable https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data.

It shouuuuld be as simple a replacement as:
```ts
const shape = frame!.shapes[(selectedShape === -1 ? 0 : selectedShape)];
if (attrname in shape) {
  const newFrame = {...frame,
      shapes: {...frame.shapes,
         [(selectedShape === -1 ? 0 : selectedShape)]:
            {...shape, properties: {...shape.properties, [attrname]: attrval } }
  }; 
}
```

Reply:
@maxkrieger `frame.shapes` is an array, so that code won't exactly work. I tried replacing it with the following code:

```
      const shapeIndex = selectedShape === -1 ? 0 : selectedShape;
      const newShapes = _.cloneDeep(frame!.shapes);
      let shape = newShapes[shapeIndex];
      shape.properties[attrname] = attrval;
      const newFrame = {
        ...frame,
        shapes: newShapes
      } as IState;
      modShapes(newFrame);
```
but it didn't want to let me cast to IState.

Reply:
> but the compiler didn't want to let me cast to IState.

Maybe check the type of `attrval` and see if that conform to `IState`'s types?

Reply:
Fixed in 0e5d3a803d6d199268633fdf5d6c1d38a92acaea